app/controllers/host_factories_controller.rb

```diff
@@ -1,34 +1,45 @@
 # frozen_string_literal: true
 
+# This controller is responsible for creating host records using
+# host factory tokens for authorization.
 class HostFactoriesController < ApplicationController
-  before_filter :verify_token,  only: :create_host
-
   # Ask the host factory to create a host.
-  # This requires the host factory's token in the Authorization header. 
+  # This requires the host factory's token in the Authorization header.
   def create_host
-    id = params.delete(:id)
-    raise ArgumentError, "id" if id.blank?
-
-    @host_builder = HostBuilder.new(@host_factory.account, id, @host_factory.role, @host_factory.role.layers, params)
-    host, api_key = @host_builder.create_host
+    host, api_key = do_create_host
     response = host.as_json
     response['api_key'] = api_key
-    
+
     render json: response, status: :created
   end
-  
+
   protected
-  
-  def verify_token
-    token = request.headers['Authorization'] or raise Unauthorized
-    if token.to_s[/^Token token="(.*)"/]
-      token = $1
-    else
-      raise Unauthorized
+
+  def do_create_host
+    host_factory = hf_token.host_factory
+    role = host_factory.role
+    HostBuilder.new(
+      host_factory.account,
+      params.require(:id),
+      role,
+      role.layers,
+      params.except(:id)
+    ).create_host
+  end
+
+  def hf_token
+    @hf_token ||= HostFactoryToken.from_token(auth_token).tap do |token|
+      raise Unauthorized unless token && [
+        token.valid?,
+        token.valid_origin?(request.ip)
+      ].all?
```

Comment:
This is neat syntax but might be simpler as a `token_valid?` helper method, like so:

```
def hf_token
  @hf_token ||= HostFactoryToken.from_token(auth_token).tap do |token|
     raise Unauthorized unless token && token_valid?(token)
  end
end

def token_valid?(token)
  token.valid? && token.valid_origin?(request.ip)
end
```

Same number of lines and it's a bit simpler. I could go either way here, though.

Reply:
Reek always complains about utility methods like that which don't touch the instance state. I'm always on the fence whether it has a point or not ;)

Reply:
@dividedmind It does have a point, but technically the solution is to create another object that accepts the param in its initializer, and while sometimes this cleans things up, sometimes the cure is worse than the disease.  My rule of thumb is having 1 or 2 "static" helper methods like that is ok... not worth creating a whole new object.

Also, there's always value in separating out validation into its own step.

All that said, your original and John's suggestion both look ok to me in this case.

Reply:
@jonahx actually I've found a good existing place for that logic, see next commit. Let me know what you think :)

Reply:
I like where you moved it.

You can remove the `or raise...` here:

```
 def hf_token
    @hf_token ||= HostFactoryToken.from_token(auth_token) or raise Unauthorized
  end
```

because the validate part handles it, and indeed you only want the `validate` one to raise.

Reply:
It does not actually :) It assumes the token isn't nil to validate.

Reply:
Yeah, I would move the nil check to the validate method then.  I don't like the repeated `raise Unauthorized`

Reply:
Reek complains about the nil check and is partly right. These are different errors (authorization missing vs. invalid); just because we report it to the client in the same way doesn't change that fact.

Reply:
I'll add a comment to that effect.